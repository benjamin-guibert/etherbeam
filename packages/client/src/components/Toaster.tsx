import React, { ReactElement } from 'react'
import Toast from './Toast'

export interface ToastContent {
  type: 'error'
  message: string
}

interface ToasterProps {
  toasts: ToastContent[]
}

const Toaster = ({ toasts }: ToasterProps): ReactElement => {
  return (
    <div className="fixed-top m-3">
      {toasts.map(({ type, message }, i) => (
        <Toast key={i} className="ml-auto" type={type}>
          {message}
        </Toast>
      ))}
    </div>
  )
}

export default Toaster
