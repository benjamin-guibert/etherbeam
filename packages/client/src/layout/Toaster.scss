@use '../styles/resources/sizes';

.my-toaster {
  flex-direction: column-reverse;
  position: fixed;
  align-items: flex-end;
  top: 0px;
  right: sizes.$spacer;
}
