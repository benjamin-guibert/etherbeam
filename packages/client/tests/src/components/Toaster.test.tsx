import * as React from 'react'
import { mount, ReactWrapper } from 'enzyme'
import Toaster from 'components/Toaster'
import { act } from 'react-dom/test-utils'

describe('<Toaster />', () => {
  beforeEach(() => {
    jest.clearAllTimers()
    jest.useFakeTimers()
  })

  let component: ReactWrapper

  describe('With toasts', () => {
    beforeAll(() => {
      component = mount(
        <Toaster
          toasts={[
            { message: 'Message 1', type: 'error' },
            { message: 'Message 2', type: 'error' },
          ]}
        />
      )
    })

    it('should have toasts', () => {
      const toasts = component.find('Toast')
      expect(toasts).toHaveLength(4)
      expect(toasts.at(0).prop('type')).toBe('error')
      expect(toasts.at(0).prop('children')).toBe('Message 1')
      expect(toasts.at(2).prop('type')).toBe('error')
      expect(toasts.at(2).prop('children')).toBe('Message 2')
    })

    it('should match snapshot', () => expect(component.render()).toMatchSnapshot())
  })

  describe('Without toasts', () => {
    beforeAll(() => {
      component = mount(<Toaster toasts={[]} />)
    })

    it('should not have toasts', () => expect(component.find('Toast')).toHaveLength(0))

    it('should match snapshot', () => expect(component.render()).toMatchSnapshot())
  })

  describe('New toast', () => {
    beforeAll(async () => {
      await act(async () => {
        const toasts = []
        component = mount(<Toaster toasts={toasts} />)
        toasts.push({ message: 'Message', type: 'error' })
        component.setProps({ toasts })
      })
    })

    it('should have toast', () => {
      const toasts = component.find('Toast')
      expect(toasts).toHaveLength(2)
      expect(toasts.at(0).prop('type')).toBe('error')
      expect(toasts.at(0).prop('children')).toBe('Message')
    })
  })
})
