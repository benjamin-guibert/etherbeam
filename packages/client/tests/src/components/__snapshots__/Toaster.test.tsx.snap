// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toaster /> With toasts should match snapshot 1`] = `
<div
  class="fixed-top m-3"
>
  <div
    aria-atomic="true"
    aria-live="assertive"
    class="fade toast ml-auto bg-danger show"
    role="alert"
  >
    <div
      class="toast-header"
    >
      <strong
        class="mr-auto"
      >
        Error
      </strong>
      <button
        class="close ml-2 mb-1"
        data-dismiss="toast"
        type="button"
      >
        <span
          aria-hidden="true"
        >
          ×
        </span>
        <span
          class="sr-only"
        >
          Close
        </span>
      </button>
    </div>
    <div
      class="toast-body"
    >
      Message 1
    </div>
  </div>
  <div
    aria-atomic="true"
    aria-live="assertive"
    class="fade toast ml-auto bg-danger show"
    role="alert"
  >
    <div
      class="toast-header"
    >
      <strong
        class="mr-auto"
      >
        Error
      </strong>
      <button
        class="close ml-2 mb-1"
        data-dismiss="toast"
        type="button"
      >
        <span
          aria-hidden="true"
        >
          ×
        </span>
        <span
          class="sr-only"
        >
          Close
        </span>
      </button>
    </div>
    <div
      class="toast-body"
    >
      Message 2
    </div>
  </div>
</div>
`;

exports[`<Toaster /> Without toasts should match snapshot 1`] = `
<div
  class="fixed-top m-3"
/>
`;
